### PR TITLE
push filter before union in qscript

### LINKED ISSUE #1381: optimization: push other operations before Union
This is similar to how we push single-side filters ahead of joins. We simply apply the same operation to both sides of the Union.

### PR DESCRIPTION
**Pull request by alissapajer**
Closes #1381.

### CODE REVIEW

File: connector/src/main/scala/quasar/qscript/Optimize.scala [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        case Map(fromInner, mf) =>
+          Map(FtoG(QC.inj(Subset(src, fromInner, sel, count))).embed, mf).some
+        case _ => None
+      }
+    case _ => None
+  }
+
+  private def filterBeforeUnion[F[_]: Functor](implicit QC: QScriptCore :<: F)
+      : QScriptCore[T[F]] => Option[QScriptCore[T[F]]] = {
+    case Filter(Embed(src), fm) =>
```
Comment (alissapajer) [inline connector/src/main/scala/quasar/qscript/Optimize.scala]: @sellout `Filter` is the only operation that seemed reasonable to push before `Union`, at least without knowing any more information about the underlying data/backend. Was there anything else you had in mind to close out #1381?

Comment (sellout) [inline connector/src/main/scala/quasar/qscript/Optimize.scala]: I think that’s probably right. If anything else comes up, we can address it then.

File: connector/src/main/scala/quasar/qscript/Optimize.scala [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  /** Should only be applied after all other QScript transformations. This gives
+    * the final, optimized QScript for conversion.
+    */
+  def optimize[F[_], G[_]: Functor](FtoG: F ~> G)(
+    implicit
+      QCF: QScriptCore :<: F,
+      QCG: QScriptCore :<: G)
+      : F[T[G]] => F[T[G]] =
+    liftFF[QScriptCore, F, T[G]](repeatedly(subsetBeforeMap[F, G](FtoG))) >>>
+      liftFF[QScriptCore, F, T[G]](repeatedly(filterBeforeUnion[G]))
```
Comment (sellout) [inline connector/src/main/scala/quasar/qscript/Optimize.scala]: Here we should probably do what we do with `coalesce` – `repeatedly(applyTransforms(subsetBeforeMap, filterBeforeUnion))`.

Comment (sellout) [inline connector/src/main/scala/quasar/qscript/Optimize.scala]: I’m not sure if the difference is testable with the current set of optimizations … you’d basically need a case where `subsetBeforeMap` could only be applied after `filterBeforeUnion` was applied. But future optimizations may work better with the `applyTransforms` versions.

**Comment by alissapajer** [comment]
@sellout ready for another look.

**Pull request merged**